onnxruntime/core/graph/graph_utils.cc

```diff
@@ -191,51 +137,221 @@ static void UpdateImplicitInputNameInSubgraph(Node& node,
+                             node.Index(),
+                             input_edge_end.GetSrcArgIndex(),
+                             input_edge_end.GetDstArgIndex(),
+                             GetNodeOutputName(input_edge_node, input_edge_end.GetSrcArgIndex())};
+
+  // Check if one of the edges provides an implicit input to a subgraph, and if so, check if the subgraph allows
+  // the node to be safely removed.
+  for (const auto& output_edge : output_edges) {
+    if (OutputEdgeProvidesImplicitInput(graph, output_edge)) {
+      Node& mutable_output_edge_node = *graph.GetNode(output_edge.dst_node);
+      if (!CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, input_edge.arg_name)) {
+        LOGS_DEFAULT(WARNING) << " Removal of node " << node.Name()
+                              << " failed due to problem in updating a subgraph.";
+        return false;
+      }
+    }
+  }
+
+  // Remove the output edges of the node and then the node itself (this will remove its input edge too).
+  RemoveOutputEdges(graph, output_edges);
+  graph.RemoveNode(node.Index());
+
+  // Create connections between the incoming node and the outgoing nodes of the node that we removed.
+  for (const auto& output_edge : output_edges) {
+    // Take care of subgraph inputs.
+    if (OutputEdgeProvidesImplicitInput(graph, output_edge)) {
+      Node& mutable_output_edge_node = *graph.GetNode(output_edge.dst_node);
+      UpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, input_edge.arg_name);
     }
+
+    // Add new edge connecting the input with the output nodes directly.
+    graph.AddEdge(input_edge.src_node, output_edge.dst_node, input_edge.src_arg_index, output_edge.dst_arg_index);
   }
 
-  // Remove output edge.
-  graph.RemoveEdge(node.Index(), output_edge_node_index, output_edge.GetSrcArgIndex(), output_edge_dst_arg);
+  return true;
+}
 
-  // Remove node (this will remove the input edge too).
+/** Remove a node with a single incoming initializer (and no other incoming node). */
+static bool RemoveNodeWithSingleInitializerIn(Graph& graph, Node& node) {
+  // Store info for input initializer and output edges.
+  auto* input_def = node.MutableInputDefs()[0];
+  std::vector<GraphEdge> output_edges = GetNodeOutputEdges(node);
+
+  // Check if one of the edges provides an implicit input to a subgraph, and if so, check if the subgraph allows
+  // the node to be safely removed.
+  for (const auto& output_edge : output_edges) {
+    if (OutputEdgeProvidesImplicitInput(graph, output_edge)) {
+      Node& mutable_output_edge_node = *graph.GetNode(output_edge.dst_node);
+      if (!CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, input_def->Name())) {
+        LOGS_DEFAULT(WARNING) << " Removal of node " << node.Name()
+                              << " failed due to problem in updating a subgraph.";
+        return false;
+      }
+    }
+  }
```

Comment:
optional: This could be a local helper function if it's the same as what happens in RemoveNodeWithSingleNodeIn.

Reply:
Done